src/backend/distributed/utils/reference_table_utils.c

```diff
@@ -22,28 +22,331 @@
+	 * We only take an access share lock, otherwise we'll hold up master_add_node.
+	 * In case of create_reference_table() and upgrade_to_reference_table(), where
+	 * we don't want concurrent writes to pg_dist_node, we have already acquired
+	 * ShareLock on pg_dist_node.
+	 */
+	List *newWorkersList = WorkersWithoutReferenceTablePlacement(shardId,
+																 AccessShareLock);
+	if (list_length(newWorkersList) == 0)
+	{
+		/* nothing to do, no need for lock */
+		UnlockColocationId(colocationId, ExclusiveLock);
+		return;
+	}
+
+	/*
+	 * master_copy_shard_placement triggers metadata sync-up, which tries to
+	 * acquire a ShareLock on pg_dist_node. We do master_copy_shad_placement
+	 * in a separate connection. If we have modified pg_dist_node in the
+	 * current backend, this will cause a deadlock.
+	 */
+	if (TransactionModifiedNodeMetadata)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot replicate reference tables in a transaction "
+							   "that modified node metadata")));
+	}
+
+	/*
+	 * Modifications to reference tables in current transaction are not visible
+	 * to master_copy_shard_placement, since it is done in a separate backend.
+	 */
+	if (AnyRelationsModifiedInTransaction(referenceTableIdList))
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot replicate reference tables in a transaction "
+							   "that modified a reference table")));
+	}
+
+	bool missingOk = false;
+	ShardPlacement *sourceShardPlacement = ActiveShardPlacement(shardId, missingOk);
+	if (sourceShardPlacement == NULL)
+	{
+		/* check for corrupt metadata */
+		ereport(ERROR, (errmsg("reference table shard "
+							   UINT64_FORMAT
+							   " does not have an active shard placement",
+							   shardId)));
+	}
+
+	/*
+	 * When using local backend for master_copy_shard_placement, current user should
+	 * own all reference table users. Otherwise master_copy_shard_placement will
+	 * error out. Superusers are regarded as owners.
+	 */
+	if (UseLocalBackendForReferenceTableReplication &&
+		!TableListOwner(referenceTableIdList))
+	{
+		ereport(ERROR, (errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+						errmsg("current user doesn't have enough permissions to "
+							   "replicate reference tables"),
+						errhint("Run 'SELECT replicate_reference_tables()' with "
+								"a superuser, and then try again.")));
+	}
+
+	WorkerNode *newWorkerNode = NULL;
+	foreach_ptr(newWorkerNode, newWorkersList)
+	{
+		ereport(NOTICE, (errmsg("replicating reference table '%s' to %s:%d ...",
+								referenceTableName, newWorkerNode->workerName,
+								newWorkerNode->workerPort)));
+
+		char transferMode = UseLocalBackendForReferenceTableReplication ?
+							TRANSFER_MODE_BLOCK_WRITES : TRANSFER_MODE_AUTOMATIC;
+
+		StringInfo placementCopyCommand =
+			CopyShardPlacementToWorkerNodeQuery(sourceShardPlacement,
+												newWorkerNode, transferMode);
+
+		if (UseLocalBackendForReferenceTableReplication)
```

Comment:
I'm still a bit unsure about the right user experience for users that can't connect to localhost.

What do you think of these options?
a) Fall back to ExecuteQueryStringIntoDestReceiver automatically if we fail to connect to localhost
b) Throw an error recommending to run replicate_reference_tables() manually
c) Keep the original behaviour (replicate_reference_tables_on_add_node) behind a GUC which defaults to off